### PR TITLE
Improve CLI upload / download experience

### PR DESCRIPTION
**Pull request by sminot**
Features added:

- Let the user select a subset of files to upload by list or glob
- Prompt the user to try again if no files were selected for download / upload
- Prompt the user to try again if the dataset was not selected correctly (if they forget to hit the tab key)
- Let the user upload hidden files

### CODE REVIEW

File: cirro/cli/interactive/download_args.py [hunk 1]
```diff
@@ -32,7 +32,17 @@ def ask_dataset(datasets: List[Dataset], input_value: str) -> str:
...
+
+    # The user has made a selection which does not match
+    # any of the options available.
+    # This is most likely because there was a typo
+    if prompt_wrapper({
+        'type': 'confirm',
+        'name': 'try_again',
+        'message': 'The selection does match an option available - try again?'
+    })['try_again']:
+        return ask_dataset(datasets)
```
Comment (nathanthorpe) [inline cirro/cli/interactive/download_args.py:44]: missing `input_value` as argument

File: cirro/cli/interactive/download_args.py [hunk 2]
```diff
@@ -72,12 +82,24 @@ def ask_dataset_files_list(files: List[File]) -> List[File]:
...
 
-    return [
+    selected_files = [
         file
         for file in files
         if file.relative_path in set(answers['files'])
     ]
 
+    if len(selected_files) == 0:
+        if prompt_wrapper({
```
Comment (nathanthorpe) [inline cirro/cli/interactive/download_args.py:92]: Could you use the ask function `ask("confirm", "No files were selected - try again?")`

File: cirro/cli/interactive/download_args.py (hunk 1 above)
Comment (sminot) [reply cirro/cli/interactive/download_args.py]: Thanks, good catch (84ae3d7)

File: cirro/cli/interactive/download_args.py (hunk 2 above)
Comment (sminot) [reply cirro/cli/interactive/download_args.py]: Much better (dcaf57d)

**Comment by sminot** [comment]
Ready for your review, @nathanthorpe

File: cirro/cli/interactive/download_args.py [hunk 3]
```diff
@@ -72,28 +81,41 @@ def ask_dataset_files_list(files: List[File]) -> List[File]:
...
+            return ask_dataset_files_list(files)
+        else:
+            raise RuntimeWarning("No files selected")
+    else:
+        return selected_files
+
 
 def ask_dataset_files_glob(files: List[File]) -> List[File]:
 
     selected_files = ask_dataset_files_glob_single(files)
```
Comment (nathanthorpe) [inline cirro/cli/interactive/download_args.py:104]: this loop could be a bit simplified as well:

```
while True:
    selected_files = ask_dataset_files_glob_single(files)
    confirmed = ask("confirm", f'Number of files selected: {len(selected_files)} / {len(files):,}')
    if confirmed:
        break
```

File: cirro/cli/interactive/upload_args.py [hunk 4]
```diff
@@ -47,61 +88,117 @@ def ask_process(processes: List[Process], input_value: str) -> str:
...
+        'message': "Include hidden files (expert: e.g. Zarr)",
+        'name': 'include_hidden',
+        'default': False
+    })['include_hidden']
+
+
+def ask_files_in_directory(data_directory) -> List[str]:
+
+    # Ask whether hidden files should be included
+    include_hidden = ask_include_hidden()
```
Comment (nathanthorpe) [inline cirro/cli/interactive/upload_args.py:103]: this can also use the `ask` function

**Review by nathanthorpe — APPROVED** [review]

File: cirro/cli/interactive/download_args.py (hunk 3 above)
Comment (sminot) [reply cirro/cli/interactive/download_args.py:104]: Similar approach, but avoiding `while True` (d58a4d3)

**Comment by sminot** [comment]
I made another change that I wanted to make sure you were ok with, @nathanthorpe 

When downloading files interactively, I'm removing the `data/` prefix when asking the user to either select from a list or apply a glob. I think this is much more intuitive to how people use the platform.

Thoughts?

**Comment by nathanthorpe** [comment]
Yeah that sounds good!

**Pull request merged**